### PR TITLE
[NPU]: Optimize xgrammar token bitmask on NPU  with AscendC

### PR DESCRIPTION
**Pull request by ChefWu551**
## Motivation

When speculative decoding is enabled and the request uses JSON Schema output format, SGLang may trigger the xgrammar backend. On NPU, the previous xgrammar token bitmask path used the torch CPU fallback, which introduces extra CPU/NPU data movement and hurts inference latency.

This PR integrates the `apply_token_bitmask` operator from `sgl-kernel-npu` to speed up this path on Ascend NPU.

Related kernel PR: https://github.com/sgl-project/sgl-kernel-npu/pull/442

## Accuracy Tests

Can refer to the PR above.

## Speed Tests and Profiling

According to the related `sgl-kernel-npu` PR, in single-concurrency scenarios:
- CPU fallback average latency: 4.48 ms
- AscendC `apply_token_bitmask` average latency: 0.171 ms

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #26617179869](https://github.com/sgl-project/sglang/actions/runs/26617179869)
Latest PR Test (Extra): :x: [Run #26617179757](https://github.com/sgl-project/sglang/actions/runs/26617179757)

### CODE REVIEW

File: python/sglang/srt/constrained/xgrammar_backend.py [hunk 1]
```diff
@@ -114,7 +115,7 @@ def apply_vocab_mask(self, logits: torch.Tensor, vocab_mask: torch.Tensor) -> No
             else:
                 apply_token_bitmask_inplace_triton(logits, vocab_mask)
         elif logits.device.type == "npu":
-            apply_token_bitmask_inplace_torch(logits, vocab_mask)
+            torch.ops.npu.apply_token_bitmask(logits, vocab_mask)
```
[the remark this answers is not in the document]
Comment (ChefWu551) [reply python/sglang/srt/constrained/xgrammar_backend.py:121]: I checked the implementation in sgl-kernel-npu PR [#442](https://github.com/sgl-project/sgl-kernel-npu/pull/442).

The operator is registered under the `npu` namespace via `TORCH_LIBRARY_FRAGMENT(npu, m)`, and the tests in that PR also call it as `torch.ops.npu.apply_token_bitmask(...)`, so `torch.ops.npu` is the intended namespace.

File: python/sglang/srt/constrained/xgrammar_backend.py [hunk 2]
```diff
@@ -35,13 +35,14 @@
...
-    apply_token_bitmask_inplace_torch,
-)
 from sglang.srt.constrained.utils import is_legacy_structural_tag
-from sglang.srt.utils import is_hip
+from sglang.srt.utils import is_hip, is_npu
 
 _is_hip = is_hip()
+_is_npu = is_npu()
+if _is_npu:
+    import sgl_kernel_npu  # noqa: F401
```
[the remark this answers is not in the document]
Comment (ChefWu551) [reply python/sglang/srt/constrained/xgrammar_backend.py]: Removed the relevant judgment logic; the corresponding package is only imported when the branch condition determines it is NPU.

**Review by DarkSharpness — APPROVED** [review]
I'm ok with that as long as it can pass ci and correctness test. BTW do we have any test in AOT kernel for this kernel? (or could you show me some of your test results)

**Comment by ChefWu551** [comment]
> I'm ok with that as long as it can pass ci and correctness test. BTW do we have any test in AOT kernel for this kernel? (or could you show me some of your test results)

Thanks! Yes, this change is based on the `apply_token_bitmask` AscendC operator that has already been added in `sgl-kernel-npu`.

For the AOT/NPU kernel side, there is a dedicated PR here: https://github.com/sgl-project/sgl-kernel-npu/pull/442

That PR includes the operator implementation, SGLang integration, validation, and benchmark results. In the reported tests, the AscendC operator was compared with the original CPU fallback path:

* Single-concurrency latency: CPU fallback 4.48 ms vs AscendC operator 0.171 ms
* 16-concurrency Qwen3-8B + Eagle3 + structured JSON output:

  * TTFT: 157.7 ms -> 107.5 ms
  * TPOT: 77.69 ms -> 37.24 ms
  * JSON parse/schema valid results stayed the same: 95/100

So this PR only wires SGLang to use the existing NPU kernel path. More details and the full test results are in the `sgl-kernel-npu` PR above.

**Comment by DarkSharpness** [comment]
cc @hnyls2002
